src/ConvertToVersionRange.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class ConvertToVersionRange : Task
+{
+    [Required]
+    public ITaskItem[] References { get; set; } = [];
+
+    [Required]
+    public string VersionProperty { get; set; } = string.Empty;
+
+    [Output]
+    public ITaskItem[] ReferencesWithVersionRanges { get; private set; } = [];
+
+    public override bool Execute()
+    {
+        var success = true;
+
+        foreach (var reference in References)
+        {
+            var automaticVersionRange = reference.GetMetadata("AutomaticVersionRange");
+
+            if (automaticVersionRange.Equals("false", StringComparison.OrdinalIgnoreCase))
+            {
+                continue;
+            }
+
+            var privateAssets = reference.GetMetadata("PrivateAssets");
+
+            if (privateAssets.Equals("All", StringComparison.OrdinalIgnoreCase))
+            {
+                continue;
+            }
+
+            var version = reference.GetMetadata(VersionProperty);
+            var match = Regex.Match(version, @"^\d+");
+
+            if (match.Value.Equals(string.Empty, StringComparison.Ordinal))
+            {
+                Log.LogError("Reference '{0}' with version '{1}' is not valid for automatic version range conversion. Fix the version or exclude the reference from conversion by setting 'AutomaticVersionRange=\"false\"' on the reference.", reference.ItemSpec, version);
+                success = false;
+                continue;
+            }
+
+            var nextMajor = Convert.ToInt32(match.Value) + 1;
+
+            var versionRange = $"[{version}, {nextMajor}.0.0)";
```

Comment:
Since we don't have to type all these anymore, is there a way to make a range that accounts for the [brokenness of NuGet version ranges](https://github.com/NuGet/Home/issues/6434) with respect to prereleases of the next major? I didn't re-read that thread again, but would `[7.0.0, 8.0.0-0)` work and is it worth it to try doing that? Or would that just result in a bunch of weird packages that would be super-odd if Microsoft ever decides to fix https://github.com/NuGet/Home/issues/6434 ?

Reply:
The last time I tried any of the workarounds listed in that thread, none of them actually worked.